### PR TITLE
[microcosm] Unify effect and domain behavior through agent class

### PR DESCRIPTION
**Pull request by nhunzaker**
This commit adds a new class to describe the dispatch behavior: an agent. Agents receive action dispatches from history and can manage their own internal state:

```javascript
import { start, stop } from '../actions/gps'

class GPS extends Agent () {
  setup() {
    this.listen()
  }

  listen() {
    this.watcher = navigator.geolocation.watchPosition(this.next)
  }

  ignore() {
    navigator.geolocation.clearWatch(this.watcher);
  }

  receive(action) {
    switch(action) {
      case start: 
        return this.listen()
      case stop: 
        return this.ignore()
      default: 
        return null
    }
  }
}

let gps = new GPS()

gps.subscribe(next => console.log(next)) // location.... location .... location....

gps.send(start)
```

I still haven't figured out the public API for Agents yet, but I think they will be a useful escape hatch for creating new types of Domain/Effect-like abstractions. I'm also curious if they could be used as a type of viewmodel class, but I'm not quite there yet.

### CODE REVIEW

File: packages/microcosm/src/subject.js [hunk 1]
```diff
@@ -6,14 +6,14 @@ import { noop, EMPTY_SUBSCRIPTION } from './empty'
 import { set, merge } from './data'
 
 export class Subject {
-  meta: { tag: *, status: string, origin: Microcosm }
+  meta: { key: *, status: string, origin: Microcosm }
```
Comment (nhunzaker) [inline packages/microcosm/src/subject.js]: I renamed `tag` to `key` for consistencies sake.

File: packages/microcosm/src/subject.js [hunk 2]
```diff
@@ -134,14 +134,14 @@ export class Subject {
   }
 
   toString(): string {
-    return this.meta.tag || 'Subject'
+    return this.meta.key
```
Comment (nhunzaker) [inline packages/microcosm/src/subject.js]: Feels good to get rid of this null state :)

File: packages/microcosm/test/helpers.js [hunk 3]
```diff
@@ -4,7 +4,7 @@ export function asTree(history) {
   function dig(focus) {
     let children = focus.children.map(dig)
 
-    return children.length ? [focus.tag, ...children] : focus.tag
+    return children.length ? [focus.key, ...children] : focus.key
   }
 
   return asciitree(dig(history.toJSON().tree))
```
Comment (nhunzaker) [inline packages/microcosm/test/helpers.js]: As a side note, this is a utility for serializing history into a tree visualization. I'm using this for snapshot testing the history structure:

```
    a
    |
    b
   / \
  c   d
```

Super useful!

File: packages/microcosm/src/agent.js [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+import { Subject } from './subject'
+import { merge } from './data'
+
+export class Agent extends Subject {
+  repo: Microcosm
+  options: Object
+
+  static defaults: ?Object
+
+  constructor(repo: Microcosm, options?: Object) {
```
Comment (nhunzaker) [inline packages/microcosm/src/agent.js]: Thinking on this a bit more, I'm also curious about the idea of an `AgentPool` class that could distribute state management across a group of presenters by sharing agents with the same configuration.

File: packages/microcosm/src/domain.js [hunk 5]
```diff
@@ -116,10 +62,22 @@ export class Domain<State: mixed = null> extends Subject {
...
+
+    if (next !== this.payload) {
+      this.next(next)
+    }
+
+    // TODO: This could probably be a generic storage solution
+    // that cleaned up keys as actions completed
+    this._ledger.clean(action)
+  }
+
```
Comment (dce) [inline packages/microcosm/src/domain.js]: ❤️ how much this cleans up Domain.

File: packages/microcosm/src/agent.js (hunk 4 above)
Comment (nhunzaker) [reply packages/microcosm/src/agent.js]: You'd do that by deriving a hashcode from the repo's identity and values passed via options

File: packages/microcosm/src/proto.js [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    Object.defineProperty(NewEntity.prototype, key, {
+      configurable: true,
+      enumerable: false,
+      writeable: true,
+      value: config[key]
+    })
+  }
+
+  return NewEntity
+}
```
Comment (dce) [inline packages/microcosm/src/proto.js]: Does this have a test?

Comment (nhunzaker) [reply packages/microcosm/src/proto.js]: It gets covered through domain/effect usage in other tests, but thanks for calling it out. I should test to cover a few cases:

- `null`
- There are certain rules for [property descriptors of ES6 classes](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/getOwnPropertyDescriptors), I should make this play nice

Really, I don't want to have something like this in Microcosm, but we've allowed users to pass domains/effects as POJOs forever. I'm thinking about deprecating it, but actually enforcing that domains/effects need to inherit from a library class might need to go out in 14.x.

File: packages/microcosm/src/agent.js [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    this.repo = repo
+    this.options = merge(repo.options, this.constructor.defaults, options)
+
+    let tracker = repo.history.subscribe(this.receive.bind(this))
+
+    repo.subscribe({
+      complete: () => {
+        tracker.unsubscribe()
+        this.teardown(this.repo, this.options)
+      }
```
Comment (efatsi) [inline packages/microcosm/src/agent.js]: looks like this is existing functionality from domains/effects, but what would cause a `complete` action?

**Comment by efatsi** [comment]
+1 to the refactoring here. Personally, I'm not too interested in a public API for Agents - I can't think of a use case for cutting off your own class that extends Agent when you have Domains/Effects at your fingertips.

**Comment by nhunzaker** [comment]
@dce Looks like I was missing coverage in a few areas. In 1e37997a I added a few more tests and added an assertion to warn users when they add a null domain/effect. This achieves complete coverage of the proto module.

**Comment by nhunzaker** [comment]
@efatsi Agreed. I don't want to advertise it for the time being, but I _am_ quite curious about how we could use it to create domains that do not rely on history.

File: packages/microcosm/src/agent.js (hunk 7 above)
Comment (nhunzaker) [reply packages/microcosm/src/agent.js]: Microcosm's are Observables (implemented as Subjects ). When you call `complete()` on a Microcosm, it has the same effect as calling `shutdown` in Microcosm 12. Instead of subscribing to the `teardown` event using an event emitter, repos/domains/presenters now listen to the `complete` lifecycle event.

TBH, I think there's an improvement that could be made here. This feels like bad design. I wonder if I should give Subject itself a setup and teardown method. You wouldn't have to implement it, but it could be inherited through Agent to Domains and Effects.

**Comment by nhunzaker** [comment]
Thanks everyone!
